### ISSUE TITLE
Highlight "New" Dashboard Notifications

### ISSUE THREAD

**Issue by stpaultim** [backdrop/backdrop-issues#5098]
**Description of the need**

In my opinion, Dashboard notifications are a pretty powerful and potentially useful tool for communicating with the community. However, it's so very easy to overlook them,especially given that they don't change very often. 

Could we do something to draw attention to NEW notifications when they first arrive?

**Proposed solution**

If a user views the dashboard and there are any new notifications. A "NEW" label in a different color is visible to draw their attention to this new notification. I think we have something like this for new content, but I don't remember which issue or how to trigger it. 

Ideally, the "New" label would disappear once a user have viewed the dashboard notification.

**Alternatives that have been considered**

This is another good idea: https://github.com/backdrop/backdrop-issues/issues/5097
The above issue triggered me to post this idea as an alternative suggestion, but I'm generally in favor of either of them - or if there should be some combination. 

**Additional information**

This issue was inspired by this problem: https://github.com/backdrop/backdrop-issues/issues/5094
We are looking for ways to make sure that users get this notification about the need to upgrade their Backup and Migrate module and some way to get a user attention while they are in the process of updating or visiting their site would be good. 

**Draft of feature description for Press Release (1 paragraph at most)**

Backdrop now includes...

**Comment by domaingood** [comment]
+1

**Comment by klonos** [comment]
> Related question: was it by intention that the notifications show without any date/time? That makes it harder for people to figure out, how current that info snippet may be.

Not intentional, but that would help. Separate issue?

**Comment by docwilmot** [comment]
> I don't think it's (technically) possible to track something like an "already read" or "already seen" status as with nodes or comments.

Cookies?

**Comment by docwilmot** [comment]
I had made [a PR that made the icon into a dropdown](https://github.com/backdrop/backdrop-issues/issues/639#issuecomment-170393205) back when we were coming up with this idea. Maybe do that? 

Or a separate method for news notifications?

**Comment by stpaultim** [comment]
I agree that a date stamp is probably a good idea. I'm a little surprise that we let that slip (I searched the original issue for any mention of that idea and did not find any). 

I also like the idea of possible using the date stamp to simply highlight any RECENT announcements (maybe within 2 weeks?). Thanks @indigoxela!

Yes, I recognize that this block is easily disabled. But, I suppose the more useful and informative it is, the less likely folks will disable it. 

I would love to get telemetry data at some point about how many sites are leaving this enabled.

**Comment by klonos** [comment]
> ...simply highlight any RECENT announcements (maybe within 2 weeks?)

Here's a PR for it: https://github.com/backdrop/backdrop/pull/3633

Adds dates to the announcements + a "new" marker to those that are less than 2 weeks old (same marker as used to denote new content in `admin/content`) + a handy `is-new` class, so people can easily target them and theme those that are new the way they like.

![image](https://user-images.githubusercontent.com/2423362/119278184-51175a80-bc2c-11eb-957f-8bf14f12c120.png)

PS: if you are testing the PR sandbox, note that the marker of that announcement as shown in the screenshot above will go away next week. So unless we add another announcement before then, there will be no way to test things in the PR sandbox after that point (announcements are hard-coded to be considered "old" after 14 days of publishing).

**Comment by klonos** [comment]
...I consider this a task rather than a new feature, so I think it can go in with the next bug fix release. Feel free to change the milestone if you think otherwise.

**Comment by indigoxela** [comment]
I like it! Displaying the date feels like a real improvement. :+1: 

The custom date format (`M j, Y`) is probably a compromise (short, yet recognizable for people used to different formats) - and the notifications are in English, anyway, so it seems OK. Works for me.

> I consider this a task rather than a new feature...

Agreed, as we only display information that already existed (timestamp), but wasn't displayed for unknown reasons.

**Comment by klonos** [comment]
> ...wasn't displayed for unknown reasons.

We just didn't think of this at the time this block was introduced 🤷🏼

**Comment by stpaultim** [comment]
I reviewed the test site and love the changes. I think that this is a real improvement. One comment, should the ALERT say "recent" instead of "new." 

If we are flagging anything in the last two weeks and the user visits the page several times in that time frame, the message won't be "new" each time. But, the message will remain "recent" regardless of how many times they see it. 

This isn't critical. I like it either way. 

Works for me too!

**Comment by klonos** [comment]
I agree with the reasoning of new vs recent, but don't the dates that are added provide a means to tell what's recent and what's not? 🤔 ...perhaps I should work on a way to mark anything "unread" as "new" ...perhaps using cookies as @docwilmot suggested.

**Comment by stpaultim** [comment]
I'm confused . You are currently checking for anything in the last two weeks, right?

For me, anything in the last two weeks is "recent", but not necessarily "new." If I've visited the site 5 times in the last week a recent item might not necessarily be new. Yes, I can tell what is recent by reading the dates, but the point of the flag is draw attention to the dates (isn't it).  

I think that the logic you are using is correct, but that simply changing the label from "new" to "recent" would be better. BUT, this is a subtle difference and I don't think it's a show stopper. I'm totally comfortable with "new" as well. 

Marking things "unread" could be good, but I got the impression that would be more work than it is worth.

**Comment by indigoxela** [comment]
> simply changing the label from "new" to "recent" would be more accurate...

Actually, I agree - "recent" seems like a better fit. But I'm fine with "new", too.

> I got the impression that (the logic) would be more work than it is worth.

I think, that's a valid point. Let's not overdo it. Displaying the dates with a small marker for a recent one is a real improvement.

**Comment by indigoxela** [comment]
> Once I've visited a new content item, the marker goes away. Could we use a similar approach here?

Probably not without (too) big effort, at least that's what @stpaultim and me think. Nodes for instance get their new/updated/read states saved in the "history" database table - which requires a node ID. ((@olafgrabienski nicht mit Kanonen auf Spatzen schießen :laughing:))

**Comment by indigoxela** [comment]
So, is there consensus that "new" is fine and the change to "recent" isn't necessary? Or do people prefer "recent"?

If "new" is fine, I'd mark this PR as RTBC. But a little feedback regarding the wording is appreciated.

**Comment by laryn** [comment]
If it's simply based on time, I'd do "recent" -- but if it's based on a cookie or some way of determining whether it's been viewed already or not, I'd do "new".

**Comment by bugfolder** [comment]
I think "new" works even for stuff 2 weeks-ε old; its use is already familiar from the Content page; and it's shorter, which is a big plus. So I vote "new."

**Comment by bugfolder** [comment]
Code looks good. Applied patch on local site, saw dates. Temporarily changed `$days_before_news_is_old` to 114 days so I could see "new" appear, which it did. LGTM.

**Comment by olafgrabienski** [comment]
I'm also fine with "new" (instead of "recent").

**Comment by stpaultim** [comment]
>> I think "new" works even for stuff 2 weeks-ε old; its use is already familiar from the Content page; and it's shorter, which is a big plus. So I vote "new."

I'm fine with either. But, just to clarify the argument for "recent" instead of "new" is that this flag DOES NOT DISAPPEAR once a user has seen it. It's not an issue of how old it is, but of how often it's been seen. 

If I've seen a post 10 times already it's not really "new" is it (but if it's less than two weeks old - it is still recent). When I see a "new" flag, this usually means NEW since I last visited the page, which will not be the case for this PR. 

Again - I'm fine with new (and this can be changed anytime). I just want to be sure the people understand the reason for the "recent" suggestion. Please, don't hold us this PR based upon this word.

**Comment by bugfolder** [comment]
@stpaultim's logic for the distinction makes sense to me, but I'd still go with "new" for now because that's the current behavior (edit a node on admin/content, and the "new" badge remains even if you refresh the page).

Revisiting the site-wide behavior (and possibly changing it) in a separate issue makes sense, too.

**Comment by stpaultim** [comment]
>> but I'd still go with "new" for now because that's the current behavior (edit a node on admin/content, and the "new" badge remains even if you refresh the page).

Ahh, I didn't realize this. I thought that this PR would be introducing a new or different behavior. I was under the impression that the "new" badge on other content was based on whether or not you had seen it before. 

If this PR is behaving the same as other examples of a "new" badge, then I would stick with the wording and keep things consistent.

**Comment by olafgrabienski** [comment]
> that's the current behavior (edit a node on admin/content, and the "new" badge remains even if you refresh the page)

Not sure if it is also the intended behavior. However, when you *view* a content item with a "New" badge and go back to `admin/content`, the badge is disappeared.

**Comment by klonos** [comment]
> ...(edit a node on admin/content, and the "new" badge remains even if you refresh the page).

Yeah, that's because you just edited the node, so the new edits you've made warrant it as new. Catch 22 situation 😅 ...as @olafgrabienski said, once you view the newly-updated node, the "new" flag goes away.

I can give the cookie option a go, and see how much "complication" it adds to the code. Gimme a few days please.

**Comment by jenlampton** [comment]
I'm really uncomfortable with using "new" here, because it looks identical to the "new" indicator on the nodes / comments listings, but means something completely different: "this was posted in the last 2 weeks" as opposed to "You've already read this post/comment." That could create a confusing experience as people struggle to find which page they need to visit to make the new markers on notifications disappear. 

I don't think we can/should use the cookie approach to determine if one of these alerts was read, because there is no secondary page to visit. For nodes and comments, there is a secondary link to visit. If someone has visited the link, we can mark the item as read. In this case, the only place this notification appears is on this same page, and I don't think we can assume that someone has read this notification block on the page if they have loaded it. I come to the dashboard all the time to access other things, and always don't read the notifications. 

What if, rather than adding a secondary indicator, we put the green highlight on the date itself? For dates in the last 2 weeks they have a green background. For those older, they don't. Because date is what determines the highlight, it would be more intuitive if we put the highlight directly onto the date.

**Comment by stpaultim** [comment]
It's seems as if there is some confusion about whether or not this badge on new notifications works differently than the new badge on posts and content. 

I've been under the impression that it does work differently (as described by @jenlampton), but @bugfolder (and I think others) is/are suggesting that the behavior is the same. 

>> "but I'd still go with "new" for now because that's the current behavior (edit a node on admin/content, and the "new" badge remains even if you refresh the page)."

I agree that some other indicator of freshness, such as a green highlight, would be an acceptable solution that avoids the word "recent." 

But, I'm not entirely sure if folks were really against the word "recent" or if they were under the impression that this works the same as the "new" content badge and wanted to keep things the same.

**Comment by indigoxela** [comment]
It seems to me that this issue is stuck on one single word. (FTR: new/recent - many people seem to be fine with either.)

How about skipping that word (badge) and only show the date - which is already helpful without any badge? Otherwise this issue might be stuck for years. That's my 2c. :wink:

**Comment by olafgrabienski** [comment]
> Do you still prefer "new" over "recent" even with the knowledge that this badge works differently from the "new" badge on content?

No preference, both have pros and cons. A green highlight on the date itself is also fine.

**Comment by stpaultim** [comment]
@klonos - This is still a candidate for inclusion in the next release. I think we have consensus on adding a date to notifications AND maybe a green highlight. Let's get a simple version of this in ASAP and make incremental changes in the future if necessary.

**Comment by klonos** [comment]
- > What if, rather than adding a secondary indicator, we put the green highlight on the date itself?

    Sure @jenlampton, here's what it would look like:

    <img width=50% src="https://user-images.githubusercontent.com/2423362/132935385-9835bcda-b6d5-4cbb-9e7d-90011601ac20.png">

    ...we could remove the background color indication from the dates, or we could have it "fade" the older the notification is (using css opacity or something like that).

- > This is still a candidate for inclusion in the next release.

    Sure @stpaultim ...but since this is a task, it could get into 1.20.1 too. So I would like to propose an alternative, in the form of a "timeline":

    <img width=50% src="https://user-images.githubusercontent.com/2423362/132934889-27747edd-ac4c-4046-b890-292fef0b6a87.png">

    I believe that it does a good job conveying the notion of new vs old notifications.

**Comment by jenlampton** [comment]
I don't know about the timeline. I've never seen a news feed with dates in the "time ago" format. I prefer the actual date.

**Comment by docwilmot** [comment]
wasnt someone working on using cookies for this?

**Comment by docwilmot** [comment]
Never mind, I scrolled a couple of lines up as I should have, and found my answer.
> I don't think we can/should use the cookie approach to determine if one of these alerts was read, because there is no secondary page to visit. 

My Gmail feed puts a `new` notification on new mail, which disappears after about 3 seconds, even if I dont read that email. Its just there to let me know that email wasnt there the last time I came to this page. I think thats a valid use for cookies.

In either case we're way overthinking this. I think https://github.com/backdrop/backdrop-issues/issues/5098#issuecomment-846633541 was already good to merge.

**Comment by stpaultim** [comment]
> I don't know about the timeline. I've never seen a news feed with dates in the "time ago" format. I prefer the actual date.

I agree. I like the dates and don't see the timeline format as an improvement.

**Comment by bugfolder** [comment]
I find the highlighted dates to be the clearest (clearer than timeline).

**Comment by klonos** [comment]
> I find the highlighted dates to be the clearest (clearer than timeline).

To clarify: do we want the dates for all notifications highlighted, or just the "recent ones"? I think that this is the last bit of feedback/decision needed here, so that I can then update the PR accordingly.

**Comment by bugfolder** [comment]
I vote:

* Show all with dates.
* But highlight only recent. (And I'm fine with any quantification of "recent".)

That way, if the user has a different idea of what time period constitutes "recent", they'll still have the dates there to decide for themselves.

But I'm fine with no-highlight or highlight-all if others prefer that.

**Comment by klonos** [comment]
Thanks @bugfolder 👍🏼 ...I'll wait for a few more people to chime in, and I will then update the PR with what is the most popular preference.

**Comment by jenlampton** [comment]
I agree with Robert. Same preference, be t also happy to go with either
other option. All are good!

**Comment by klonos** [comment]
Final thought: the current PR marks any notification that is less than 14 days old as "new". Would it be an acceptable solution to add a new timestamp field in the `[users]` table, which tracks the last time the user has seen the "Backdrop News" block (similar to how we track user creation and last login timestamp), and then use that timestamp to determine which notifications are new (which ones the user is seeing for the first time)?

I have a local copy which works like so:
1. when a user logs in for the first time and they visit the Dashboard, all notifications are marked as "new"
2. if the user refreshes the Dashboard page (or any page that holds a "Backdrop News" block for that matter), the "new" indicators go away.
3. subsequent logins and visits to the page that holds the block do not show any "new" indicators, so long as the same set of notification is present in the block
4. soon as any new notification is published and shown in the Dashboard block, that notification will be marked as "new" when the user sees the block again
5. soon as the user refreshes the page, or if they navigate away from it and then return back to it, the "new" marker goes away ...and so on.

Caveats:
- I understand that in this logic the absence of the "new" indicator does not mean that the person has actually clicked the links in the notifications and read them - it only means that these notifications were there the last time the person saw that page/block. ...at the same time, we do not have the opposite problem, where people see the "new" notifications, click on the links and read them, yet they are still marked as "new" next time they see the block.
- If the "Backdrop News" block is placed at the bottom of a long page, and the user never actually scrolls down, then the "new" indicators will go away regardless.

Thoughts?

**Comment by klonos** [comment]
PR if people wanna compare the solution I described in my previous comment: https://github.com/backdrop/backdrop/pull/3736

And here if the code diff to the first/previous PR: [`77e4bec` (#3736)](https://github.com/backdrop/backdrop/pull/3736/commits/77e4beca04fd00d488d7d06b8801ce9d9c5878bd)

**Comment by bugfolder** [comment]
I've tried it out. This looks like a nice behavior. And there's always the dates present to resolve any newness ambiguity.

**Comment by jenlampton** [comment]
I don't like the addition of the database entry for that same reasons
previously discussed.

**Comment by stpaultim** [comment]
@klonos - Sorry to report, but I don't think your new PR is working. If I use your steps to test it works as expected. When I create a new account and login, all notifications are marked as "NEW" the first time I view the dashboard, then when I revisit the page the "NEW" message is gone. However, if I add a new notification on BackdropCMS.org and then wait for it to appear (due to caching, I have to wait quite a while), it does not have the "New" tag when it finally does appear. 

Truthfully, even if this works, I'm not sure it's an improvement. I was entirely happy highlighting recent notifications for two weeks (or a week). Replacing that behavior with a one-time "New" flag that the user may or may not see isn't necessarily an improvement in my view. However, I see how it might be considered more consistent with other similar behaviors. Also, simply having a date listed is an improvement over the current situation, so I'm good with this PR if we can fix the current problem with the "New" tag not fully working as expected. 

If fixing the current problem is going to delay this, then I would rather we go back to prior versions that did work and accomplished our original goals.

**Comment by jenlampton** [comment]
I also still don't like using the new indicator for this purpose :/

**Comment by klonos** [comment]
Fair enough @stpaultim and @jenlampton ...I've closed the alternative PR, and I plan to update the previous PR to:
1. remove the "new" indicator
2. highlight the dates of recent notifications (less than 14 days)

Is that what we agree upon, or simply add the dates and do not provide any highlight/indicator at all?

**Comment by stpaultim** [comment]
> Is that what we agree upon, or simply add the dates and do not provide any highlight/indicator at all?

I went back through the issue and found that (in my reading) @olafgrabienski, @indigoxela, @jenlampton, @bugfolder and myself have all explicitly indicated approval of the idea of "highlighting" recent notifications. @docwilmot seemed to generally agree that previous suggested approaches were fine without explicitly endorsing this one. @laryn and @domaingood haven't commented since the Date/Indicator compromise was suggested. 
I don't see anyone advocating that we NOT highlight recent dates. 

If anyone would like to clarify their opinion, please jump in.

**Comment by indigoxela** [comment]
I'm a friend of simple solutions. Displaying the date has - to my understanding - wide support here. Some sort of highlighting has good support, too. But it seems to me, we all got carried away in figuring out, how that highlighting should happen.

A decent "new" or "recent" marker seems fine to me. I'm still happy with either of the words and am also happy with just visually highlighting the date.

What I *don't* consider a good idea is adding database columns (or properties to user objects), additional js (or worse, an additional js library) for such a tiny task. Also, completely changing the news display seems unnecessary to me.

So, we have consensus that displaying the dates makes sense. If we fail to find consensus on how to highlight recent news, can we discuss it in a follow-up issue? Let's not let a badge prevent a simple improvement. :wink:

**Comment by klonos** [comment]
Agreed 👍🏼 ...PR updated accordingly:

<img width=70% src="https://user-images.githubusercontent.com/2423362/133119889-5d273eaf-6c0d-4849-b04e-d7784d6c509d.png">

Tests are green!!! 🎉 🎉 🎉

**Comment by jenlampton** [comment]
@klonos I left one small comment on the PR about choosing a better variable name. Otherwise looks good!

**Comment by klonos** [comment]
Thanks @jenlampton ...naming things is hard 😅 ...can you please see my counter-suggestion in the PR comments?

**Comment by jenlampton** [comment]
I love it! If you commit, I can RTBC :)

**Comment by klonos** [comment]
Done [`f57e8a0` (#3633)](https://github.com/backdrop/backdrop/pull/3633/commits/f57e8a0be11a952070cb749eac46a8d30e621d6f)

**Comment by stpaultim** [comment]
@quicksketch - Sorry, I guess I was wrong. This never was RTBC. (I just raised this briefly in dev meeting today).

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/3633 into 1.x and 1.20.x. Thanks folks!
